torchprime/torch_xla_models/model/llama4/model.py

```diff
@@ -136,44 +136,105 @@ def __init__(self, config):
   @xp.trace_me("Llama4TextMoe")
-  def forward(self, hidden_states):
+  def forward_original(self, hidden_states):
+    # batch, seq_len, hidden_dim is the shape of hidden states
     batch, seq_len, hidden_dim = hidden_states.shape
     hidden_states = hidden_states.view(-1, self.hidden_dim)
+    # Get routed logits for each token
+    # Dimensions: batch * seq_len, num_experts
     router_logits = self.router(hidden_states)
-    tokens_per_expert = batch * seq_len
 
+    # Find the top K experts and their scores for each token
+    # router_top_value, router_indices: [batch * seq_len, top_k]
     router_top_value, router_indices = torch.topk(router_logits, self.top_k, dim=1)
+
+    # Create a score mask to keep only the top-k scores
+    # Dimensions: [num_experts, batch * seq_len]
     router_scores = (
       torch.full_like(router_logits, float("-inf"))
       .scatter_(1, router_indices, router_top_value)
       .transpose(0, 1)
     )
     # We do this to make sure we have -inf for non topK tokens before going through the !
     # Here we are just creating a tensor to index each and every single one of the hidden states. Let s maybe register a buffer for this!
-    router_indices = (
-      torch.arange(tokens_per_expert, device=hidden_states.device)
-      .view(1, -1)
-      .expand(router_scores.size(0), -1)
-    )
     router_scores = torch.sigmoid(router_scores.float()).to(hidden_states.dtype)
 
-    router_indices = router_indices.reshape(-1, 1).expand(-1, hidden_dim)
-    routed_in = torch.gather(
-      input=hidden_states,
-      dim=0,
-      index=router_indices,
-    ).to(hidden_states.device)
-    # we gather inputs corresponding to each expert based on the router indices
+    # Repeating hidden_states for num_experts
+    # This is a huge tensor, dimensions would be:
+    # num_experts, batch, seq_len, hidden_dim
+    routed_in = hidden_states.repeat(self.num_experts, 1)
+    # Gather inputs corresponding to each expert based on the router indices.
+    # Following line would likely OOM on materialization of routed_in tensor.
     routed_in = routed_in * router_scores.reshape(-1, 1)
     routed_out = self.experts(routed_in)
     out = self.shared_expert(hidden_states)
-    # now that we finished expert computation -> we scatter add because we gathered previously
-    # we have to do this because we used all experts on all tokens. This is faster than the for loop, tho you are compute bound
-    # this scales a lot better if you do EP!
+    # Now that we finished expert computation -> we scatter add because we gathered previously.
+    # W.e have to do this because we used all experts on all tokens.
     out.scatter_add_(dim=0, index=router_indices, src=routed_out.view(-1, hidden_dim))
     return out, router_scores
 
+  @xp.trace_me("Llama4TextMoe")
+  def forward(self, hidden_states):
+    """
+    This version processes experts sequentially to save memory. It is slower
```

Comment:
is the real solution to enable expert parallelism? Also if you keep increasing tensor parallelism dimension, it still OOM?

Reply:
just curious, if you make tensor parallelism to 16. can it work?

Reply:
I think I tried that, and it still didn't work for some reason. Didn't try to debug why.